### PR TITLE
feat(new schema): Redesign local relationship query filter

### PR DESCRIPTION
**Pull request by zhixuanjia**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

**Review by jsdonn — COMMENTED** [review]
Could you please share some context on why you are refactoring the class to not extend BaseQueryDAO?

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/utils/SQLStatementUtils.java [hunk 1]
```diff
@@ -257,30 +260,32 @@ public static String whereClause(@Nonnull Map<Condition, String> supportedCondit
...
-      for (Pair<Condition, String> pair : entry.getValue()) {
-        orClauses.add(entry.getKey() + supportedCondition.get(pair.getValue0()) + "'" + pair.getValue1() + "'");
+      for (Pair<Condition, LocalRelationshipValue> pair : entry.getValue()) {
+        if (pair.getValue0() == Condition.IN) {
+          if (!pair.getValue1().isArray()) {
+            throw new IllegalArgumentException("IN condition must be paired with array value");
+          }
+          orClauses.add(entry.getKey() + " IN " +  parseLocalRelationshipValue(pair.getValue1()));
+        } else {
+          orClauses.add(entry.getKey() + supportedCondition.get(pair.getValue0()) + "'" + parseLocalRelationshipValue(pair.getValue1()) + "'");
```
Comment (jsdonn) [inline dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/utils/SQLStatementUtils.java:287]: What about CONTAIN? Is that supported by local relationship filter? Do you need to add a special case for CONTAIN?

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/utils/SQLStatementUtils.java [hunk 2]
```diff
@@ -300,4 +305,36 @@ public static String whereClause(@Nonnull Filter filter, @Nonnull Map<Condition,
 
     return String.join(" AND ", andClauses);
   }
+
+  private static String parseLocalRelationshipField(@Nonnull final LocalRelationshipCriterion localRelationshipCriterion, @Nullable String tablePrefix) {
+    tablePrefix = tablePrefix == null ? "" : tablePrefix + ".";
```
Comment (jsdonn) [inline dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/utils/SQLStatementUtils.java:310]: Is there a case where tablePrefix can be ""?

File: dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java [hunk 3]
```diff
@@ -148,10 +182,23 @@ public void testFindEntitiesOneHopAwayIncomingDirection() throws Exception {
...
     // Find all Alice's direct reports.
-    Criterion filterCriterion = new Criterion().setField("i_aspectfoo$value").setValue("Alice").setCondition(Condition.EQUAL);
-    Filter filter = new Filter().setCriteria(new CriterionArray(filterCriterion));
-    List<RecordTemplate> aliceDirectReports = _localRelationshipQueryDAO.findEntities(FooSnapshot.class, new Filter().setCriteria(new CriterionArray()),
-        FooSnapshot.class, filter, ReportsTo.class, new RelationshipFilter().setDirection(RelationshipDirection.INCOMING).setCriteria(new CriterionArray()),
+    AspectField aspectField = new AspectField().setAspect(AspectFoo.class.getCanonicalName()).setPath("/value");
+    LocalRelationshipCriterion.Field field = new LocalRelationshipCriterion.Field();
+    field.setAspectField(aspectField);
+
+    LocalRelationshipCriterion filterCrtierion = new LocalRelationshipCriterion()
```
Comment (jsdonn) [inline dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java:189]: nit: typo - filterCriterion

File: dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java [hunk 4]
```diff
@@ -116,10 +141,19 @@ public void testFindOneRelationship() throws Exception {
...
     // Find all reports-to relationship for Alice.
-    Criterion filterCriterion = new Criterion().setField("i_aspectfoo$value").setValue("Alice").setCondition(Condition.EQUAL);
-    Filter filter = new Filter().setCriteria(new CriterionArray(filterCriterion));
-    List<ReportsTo> reportsToAlice = _localRelationshipQueryDAO.findRelationships(FooSnapshot.class, new Filter().setCriteria(new CriterionArray()),
-        FooSnapshot.class, filter, ReportsTo.class, new Filter().setCriteria(new CriterionArray()), 0, 10);
+    AspectField aspectField = new AspectField().setAspect(AspectFoo.class.getCanonicalName()).setPath("/value");
+    LocalRelationshipCriterion.Field field = new LocalRelationshipCriterion.Field();
+    field.setAspectField(aspectField);
+
+    LocalRelationshipCriterion filterCrtierion = new LocalRelationshipCriterion()
```
Comment (jsdonn) [inline dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java:148]: nit: typo - filterCriterion

File: dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java [hunk 5]
```diff
@@ -193,13 +240,26 @@ public void testFindEntitiesOneHopAwayOutgoingDirection() throws Exception {
     _localRelationshipWriterDAO.addRelationship(bobBelongsToStandford);
 
     // Alice filter
-    Criterion aliceFilterCriterion = new Criterion().setField("i_aspectfoo$value").setValue("Alice").setCondition(Condition.EQUAL);
-    Filter aliceFilter = new Filter().setCriteria(new CriterionArray(aliceFilterCriterion));
+    AspectField aspectField = new AspectField().setAspect(AspectFoo.class.getCanonicalName()).setPath("/value");
+    LocalRelationshipCriterion.Field field = new LocalRelationshipCriterion.Field();
+    field.setAspectField(aspectField);
+
+    LocalRelationshipCriterion filterCrtierion = new LocalRelationshipCriterion()
```
Comment (jsdonn) [inline dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java:247]: nit: typo

File: dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java [hunk 6]
```diff
@@ -245,17 +305,39 @@ public void testFindEntitiesOneHopAwayUndirected() throws Exception {
     _localRelationshipWriterDAO.addRelationship(alicePairsWithJohn);
 
     // Alice filter
-    Criterion aliceFilterCriterion = new Criterion().setField("i_aspectfoo$value").setValue("Alice").setCondition(Condition.EQUAL);
-    Filter aliceFilter = new Filter().setCriteria(new CriterionArray(aliceFilterCriterion));
+    AspectField aspectField = new AspectField().setAspect(AspectFoo.class.getCanonicalName()).setPath("/value");
+    LocalRelationshipCriterion.Field field = new LocalRelationshipCriterion.Field();
+    field.setAspectField(aspectField);
+
+    LocalRelationshipCriterion filterCrtierion = new LocalRelationshipCriterion()
```
Comment (jsdonn) [inline dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java:312]: nit: typo

File: dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java [hunk 7]
```diff
@@ -245,17 +305,39 @@ public void testFindEntitiesOneHopAwayUndirected() throws Exception {
...
+    LocalRelationshipFilter aliceFilter = new LocalRelationshipFilter().setCriteria(new LocalRelationshipCriterionArray(filterCrtierion));
 
     // Age filter
-    Criterion ageFilterCriterion = new Criterion().setField("i_aspectbar$value").setValue("30").setCondition(Condition.GREATER_THAN);
-    Filter ageFilter = new Filter().setCriteria(new CriterionArray(ageFilterCriterion));
+    AspectField aspectField2 = new AspectField().setAspect(AspectBar.class.getCanonicalName()).setPath("/value");
+    LocalRelationshipCriterion.Field field2 = new LocalRelationshipCriterion.Field();
+    field2.setAspectField(aspectField2);
+
+    LocalRelationshipCriterion filterCrtierion2 = new LocalRelationshipCriterion()
```
Comment (jsdonn) [inline dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java:324]: nit: typo

File: dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java [hunk 8]
```diff
@@ -269,4 +351,27 @@ public void testFindEntitiesOneHopAwayUndirected() throws Exception {
...
+  public void testFindOneEntityWithInCondition() throws URISyntaxException {
+    // Ingest data
+    _fooUrnEBeanLocalAccess.add(new FooUrn(1), new AspectFoo().setValue("foo"), AspectFoo.class, new AuditStamp());
+
+    // Prepare filter
+    AspectField aspectField = new AspectField().setAspect(AspectFoo.class.getCanonicalName()).setPath("/value");
+    LocalRelationshipCriterion.Field field = new LocalRelationshipCriterion.Field();
+    field.setAspectField(aspectField);
+
+    LocalRelationshipCriterion filterCrtierion = new LocalRelationshipCriterion()
```
Comment (jsdonn) [inline dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java:365]: nit: typo

File: dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java [hunk 9]
```diff
@@ -269,4 +351,27 @@ public void testFindEntitiesOneHopAwayUndirected() throws Exception {
     Set<FooUrn> expected = ImmutableSet.of(bob, john);
     assertEquals(actual, expected);
   }
+
+  @Test
+  public void testFindOneEntityWithInCondition() throws URISyntaxException {
```
Comment (jsdonn) [inline dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java:356]: want to add a test case for opposite? aka no entities match the IN filter

**Review by jsdonn — CHANGES_REQUESTED** [review]

File: dao-api/src/main/pegasus/com/linkedin/metadata/query/UrnField.pdl [hunk 10]
```diff
@@ -0,0 +1,5 @@
+namespace com.linkedin.metadata.query
+
+record UrnField {
```
Comment (jsdonn) [inline dao-api/src/main/pegasus/com/linkedin/metadata/query/UrnField.pdl:3]: So are all of these *Field models going to be general or will users need to onboard their specific use cases?

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/utils/SQLStatementUtils.java (hunk 1 above)
Comment (zhixuanjia) [reply dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/utils/SQLStatementUtils.java:287]: So CONTAINS is not supported for now. https://github.com/linkedin/datahub-gma/blob/master/dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipQueryDAO.java#L48

But if we support it in the future, we will revisit this part of code.

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/utils/SQLStatementUtils.java (hunk 2 above)
Comment (zhixuanjia) [reply dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/utils/SQLStatementUtils.java:310]: If there is only one table involved in the SQL query, then table prefix is not needed. For example, the findEntities api https://github.com/linkedin/datahub-gma/blob/master/dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipQueryDAO.java#L48 only needs to query one entity table so no need to have table prefix.

File: dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java (hunk 3 above)
Comment (zhixuanjia) [reply dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java:189]: updated.

File: dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java (hunk 4 above)
Comment (zhixuanjia) [reply dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java:148]: updated

File: dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java (hunk 5 above)
Comment (zhixuanjia) [reply dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java:247]: updated

File: dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java (hunk 6 above)
Comment (zhixuanjia) [reply dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java:312]: updated

File: dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java (hunk 7 above)
Comment (zhixuanjia) [reply dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java:324]: updated

File: dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java (hunk 8 above)
Comment (zhixuanjia) [reply dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java:365]: updated

File: dao-api/src/main/pegasus/com/linkedin/metadata/query/UrnField.pdl (hunk 10 above)
Comment (zhixuanjia) [reply dao-api/src/main/pegasus/com/linkedin/metadata/query/UrnField.pdl:3]: So I wrote a doc to explain why we have these *Field. https://docs.google.com/document/d/1U77W5OeubBvSZPkhouMLD7Q99vwsGoUmH9iri12v_GQ/edit#

File: dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java (hunk 9 above)
Comment (zhixuanjia) [reply dao-impl/ebean-dao/src/test/java/com/linkedin/metadata/dao/localrelationship/EbeanLocalRelationshipQueryDAOTest.java:356]: sure, added a test for no match case

**Comment by zhixuanjia** [comment]
This PR has two main purposes. 1. Properly support IN condition. 2. Replace the Filter with LocalRelationshipFilter. The Filter used in BaseQueryDAO only has minimal capability. For example, the only value type it supports is string and user can only pass in string as field as well.

File: dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipQueryDAO.java [hunk 11]
```diff
@@ -217,23 +184,22 @@ private <ENTITY extends RecordTemplate> void validateEntityFilter(@Nonnull Filte
    * 2. Relationship direction cannot be unknown.
    * If any of above is violated, throw IllegalArgumentException.
    */
-  private void validateRelationshipFilter(@Nonnull RelationshipFilter filter) {
+  private void validateRelationshipFilter(@Nonnull LocalRelationshipFilter filter) {
     if (filter.getDirection() == RelationshipDirection.$UNKNOWN) {
```
Comment (jsdonn) [inline dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipQueryDAO.java:188]: Why is the relationship direction included in the relationship filter? What happens if we use an empty filter when querying? I didn't see a test case for that.

Comment (zhixuanjia) [reply dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipQueryDAO.java:188]: So the direction in the relationship filter is used to [generate the SQL](https://github.com/linkedin/datahub-gma/blob/0d89f5f62187959ce8884d02b0d93417853a4a3e/dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/utils/MultiHopsTraversalSqlGenerator.java#L47) for querying relationships.

Comment (zhixuanjia) [reply dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipQueryDAO.java:188]: It is also used to determine [what should be returned](https://github.com/linkedin/datahub-gma/blob/master/dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipQueryDAO.java#L108).

Comment (jsdonn) [reply dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipQueryDAO.java:188]: Cool, thanks for sharing. Could you also add a test case where you use the EMPTY_FILTER?

Comment (zhixuanjia) [reply dao-impl/ebean-dao/src/main/java/com/linkedin/metadata/dao/EbeanLocalRelationshipQueryDAO.java]: Sure, added a test case: `testFindEntitiesWithEmptyRelationshipFilter`

**Review by jsdonn — APPROVED** [review]

**Pull request merged**